File: library/src/commonMain/kotlin/converters/Converter.kt
package converters

import readers.Concept
import readers.Transaction
import writers.*

class Converter(private val timestampConverter: TimestampConverter) {
    fun convert(
        concepts: Sequence<Concept>,
        commentTags: Set<String>,
        idPrefix: String?,
    ): Sequence<TermEntry> {
        return concepts.map { c ->
            TermEntry(
                listOf(
                    Descrip(
                        "conceptId",
                        when(idPrefix) {
                            is String -> "$idPrefix-${c.id}"
                            else -> "${c.id}"
                        }
                    )
                ),
                c.languages.map { lang ->
                    LangSet(lang.lang, lang.terms.map { term ->
                        Term(term.value,
                            convertComment(term, commentTags),
                            term.transactions
                                .flatMap(this::convertTransaction))
                    })
                },
            )
        }
    }

    private fun convertComment(term: readers.Term, commentTags: Set<String>): String? {
        val commentElements = term.descrip.asIterable()
            .filter { it.key in commentTags }
        if (commentElements.size > 1)
            error("Multiple comments elements for term ${term.value}")
        return commentElements.singleOrNull()?.value
    }

    private fun convertTransaction(it: Transaction): List<TermNote> {
        return when (it) {
            is Transaction.Origination -> listOf(
                TermNote("createdBy", it.author),
                TermNote("createdAt", convertDate(it.date)),
            )
            is Transaction.Modification -> listOf(
                TermNote("lastModifiedBy", it.author),
                TermNote("lastModifiedAt", convertDate(it.date)),
            )
        }
    }

    private fun convertDate(date: String): String {
        val timestamp = timestampConverter.timestampFromIsoDate(date)
        val millisecondTimestamp = timestamp * 1000
        return millisecondTimestamp.toString()
    }
}

interface TimestampConverter {
    fun timestampFromIsoDate(date: String): Long
}

File: tbxconverter/converter.py
import sys
from datetime import datetime, timezone
from typing import Iterable

from .trados_reader import Concept, TransactionType
from .trados_reader import read_trados_file
from .tbx_writer import TermEntry, Term, TermNote, LangSet, Descrip
from .tbx_writer import write_tbx_file

ORIGINATION = TransactionType.ORIGINATION
MODIFICATION = TransactionType.MODIFICATION


def trados_to_tbx(
        trados_data: Iterable[Concept],
        id_prefix=None,
) -> Iterable[TermEntry]:

    def convert_terms(terms):
        return [Term(term.text,
                     convert_comment(term.descriptions),
                     [note for t in term.transactions
                      for note in convert_transaction(t)])
                for term in terms]

    def convert_comment(descriptions):
        if len(descriptions) > 1:
            return "\n".join(f"{k}: {v}" for k, v in descriptions.items())
        elif not descriptions:
            return None
        else:
            return next(iter(descriptions.values()))

    def convert_transaction(transaction):
        if transaction.type == ORIGINATION:
            return [TermNote("createdBy", transaction.source),
                    TermNote("createdAt", convert_date(transaction.date))]
        else:
            return [TermNote("lastModifiedBy", transaction.source),
                    TermNote("lastModifiedAt", convert_date(transaction.date))]

    def convert_date(date_str):
        d = datetime.fromisoformat(date_str)
        if d.tzinfo is None:
            d = d.replace(tzinfo=timezone.utc)
        return str(int(d.timestamp()) * 1000)

    for concept in trados_data:
        tbx_concept_id = (f"{id_prefix}-{concept.id}"
                          if id_prefix else str(concept.id))
        descrips = [Descrip("conceptId", tbx_concept_id)]
        languages = [LangSet(lang.code, convert_terms(lang.terms))
                     for lang in concept.languages]

        yield TermEntry(descrips, languages)


def main():
    with open(sys.argv[1]) as f:
        trados_data = read_trados_file(f)
        tbx_data = trados_to_tbx(trados_data)
        write_tbx_file(sys.stdout.buffer, tbx_data)


if __name__ == "__main__":
    main()
